src/providers/ogr/qgsogrexpressioncompiler.cpp

```diff
@@ -43,6 +43,8 @@ QgsSqlExpressionCompiler::Result QgsOgrExpressionCompiler::compile( const QgsExp
     return Fail;
   else if ( mSource->mDriverName == "MSSQLSpatial" )
     return Fail;
+  else if ( mSource->mDriverName == "GPKG" )
```

Comment:
My understanding was that OGR_L_SetAttributeFilter will transparently hand the filter off to the driver for those special cases, so there's likely to be a difference in how the driver's provider handles the expression vs how the OGR_L_SetAttributeFilter is handled for the other drivers. That was my motivation for originally failing in these cases (and given that at the time they were all already handled by dedicated providers it didn't seem like an issue).